examples/applications/object_detection/retina_net/basic/pascal_voc/loss.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+
+class FocalLoss(tf.losses.Loss):
+    def __init__(self, num_classes, alpha=0.25, gamma=2.0, delta=1.0, **kwargs):
+        super().__init__(**kwargs)
+        self.num_classes = num_classes
+        self.classification_loss = RetinaNetClassificationLoss(
+            alpha=alpha,
+            gamma=gamma,
+        )
+        self.box_loss = RetinaNetBoxLoss(delta=delta)
+
+    def call(self, y_true, y_pred):
+        box_labels = y_true[:, :, :4]
+        box_predictions = y_pred[:, :, :4]
+
+        cls_labels = tf.one_hot(
+            tf.cast(y_true[:, :, 4], dtype=tf.int32),
+            depth=self.num_classes,
+            dtype=tf.float32,
+        )
+        cls_predictions = y_pred[:, :, 4:]
+        positive_mask = tf.cast(tf.greater(y_true[:, :, 4], -1.0), dtype=tf.float32)
+        ignore_mask = tf.cast(tf.equal(y_true[:, :, 4], -2.0), dtype=tf.float32)
+
+        classification_loss = self.classification_loss(cls_labels, cls_predictions)
+        box_loss = self.box_loss(box_labels, box_predictions)
+
+        classification_loss = tf.where(
+            tf.equal(ignore_mask, 1.0), 0.0, classification_loss
+        )
+        box_loss = tf.where(tf.equal(positive_mask, 1.0), box_loss, 0.0)
+        normalizer = tf.reduce_sum(positive_mask, axis=-1)
+        classification_loss = tf.math.divide_no_nan(
+            tf.reduce_sum(classification_loss, axis=-1), normalizer
+        )
+        box_loss = tf.math.divide_no_nan(tf.reduce_sum(box_loss, axis=-1), normalizer)
+
+        classification_loss = tf.math.reduce_sum(classification_loss, axis=-1)
+        box_loss = tf.math.reduce_sum(box_loss, axis=-1)
+        return (classification_loss + box_loss, classification_loss, box_loss)
```

Comment:
btw, does it return per-example loss or reduce loss (scalar). Keras might need to per-example loss, so that it might apply sample weights to it, before final aggregation.

Reply:
lets repro this on https://github.com/keras-team/keras-cv/pull/615